### PR TITLE
fix(create-cloudflare): update next.js templates + support, use next-on-pages v1

### PR DESCRIPTION
**Pull request by james-elicx**
**What this PR solves / how to test:**

- Bumps Next.js to v13.4.2 (app directory was made stable in v13.4).
- Proper support for app directory route handlers.
  - Next.js' create tool doesn't include API routes anymore in app dir, so this creates one if one didn't already exist.
  - Uses new route handler templates when in the app directory.
- Remove `next.config.js` edge runtime option since that is no longer supported.
- Ties `@cloudflare/next-on-pages` to v1.

App deployed using the changes: https://orange-art-226f.pages.dev/api/hello

**Associated docs issue(s)/PR(s):**

- [insert associated docs issue(s)/PR(s)]

**Author has included the following, where applicable:**

- [ ] Tests
- [x] Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))

**Reviewer is to perform the following, as applicable:**

- Checked for inclusion of relevant tests
- Checked for inclusion of a relevant changeset
- Checked for creation of associated docs updates
- Manually pulled down the changes and spot-tested

### CODE REVIEW

File: packages/create-cloudflare/src/frameworks/next/index.ts
Comment (dario-piotrowicz) [inline packages/create-cloudflare/src/frameworks/next/index.ts:34]: I am not the biggest fan of returning tuples, sorry I am not too familiar with this codebase, is it common here?

File: packages/create-cloudflare/src/frameworks/next/index.ts
Comment (dario-piotrowicz) [inline packages/create-cloudflare/src/frameworks/next/index.ts:66]: If you're not checking for an `api` directory I think you need to update accordingly the variable name and the error message, no?

**Comment by dario-piotrowicz** [comment]
We have a bit of an issue here, I have `wrangler@2.20.0` installed globally, since the newly created Next apps don't add a wrangler dependency and the latest next-on-pages runs only on `wrangler@3` if I try running my Next app and I try accessing `/api/hello` I get an internal server error:

Screenshot 2023-05-18 at 11 25 54

I think this needs to be generally fixed and is probably not something that needs to be done in this PR, but it might or might not be ok to merge this PR with such issue, up to @jculvey / @petebacondarwin 🙂

File: packages/create-cloudflare/src/frameworks/next/index.ts
Comment (james-elicx) [reply packages/create-cloudflare/src/frameworks/next/index.ts:34]: The previous version did it with a ternary as it assumed there was only one type of template that could be used, but since there are two types, I moved the logic deciding which path+template to use to its own function. When doing this though, I kept the same way that was used previously - using an array that gets destructured.

**Comment by dario-piotrowicz** [comment]
Actually the above made me remind that we need to update the next-on-pages wrangler peer-dependency 😓 
https://github.com/cloudflare/next-on-pages/pull/259

File: packages/create-cloudflare/src/frameworks/next/index.ts
Comment (james-elicx) [reply packages/create-cloudflare/src/frameworks/next/index.ts]: modified the message and variable name

File: packages/create-cloudflare/src/frameworks/next/index.ts
Comment (dario-piotrowicz) [reply packages/create-cloudflare/src/frameworks/next/index.ts:34]: but before it was within a function and with well named variables 🤔 

now it's a return value

as I said I am not the biggest fan (as I personally find it more difficult to read and more error-prone) but I am ok with it if others are 🙂

**Review by dario-piotrowicz — APPROVED** [review]
Looks good to me, also I've tested it locally with an app dir application and a pages one and all seems to be working nicely 😃 

(except for the wrangler issue I mentioned in my comment, which is sort of unrelated?)

**Comment by james-elicx** [comment]
> Code changes look good! Thanks for the PR. I wanted to give an update as to not leave you hanging.
> 
> The unfortunate thing about bumping to the newest version of next and next-on-pages is that it will require the nodejs_compat flag to be set in order to build and run properly. This means that for the c3 workflow, it'll always make the user's first deployment fail, and require the to visit the dashboard to set the compat flag to proceed.
> 
> We're in the midst of an internal discussion to see if there's a way we can fix this before changing versions.

Thanks for letting me know.

Yeah, it's unfortunate. It would be nice if there was a way to prefill the compatibility flags for the user when they're creating a new Next.js project (through the preset or through wrangler), but I imagine that would probably require some significant changes to how creating a new Pages project works

**Comment by jculvey** [comment]
@james-elicx update: Just merged [a PR](https://github.com/cloudflare/workers-sdk/pull/3317) that adds the `--compatibility-flag` option to `wrangler pages project create`. 

This should unblock the PR but we'll need to update it to take advantage of the new flag. I'm thinking we can just add an optional `compatibilityFlags` entry to the `FrameworkConfig` that will get passed along in `createPagesProject`. Happy to take this part of it on if you don't have cycles.

**Comment by james-elicx** [comment]
> @james-elicx update: Just merged [a PR](https://github.com/cloudflare/workers-sdk/pull/3317) that adds the `--compatibility-flag` option to `wrangler pages project create`.
> 
> This should unblock the PR but we'll need to update it to take advantage of the new flag. I'm thinking we can just add an optional `compatibilityFlags` entry to the `FrameworkConfig` that will get passed along in `createPagesProject`. Happy to take this part of it on if you don't have cycles.

That's great! I've just pushed a commit for it.

Tested changing it to use wrangler@beta and had success creating a new Next.js project with C3 and the compat flags 🙂

**Review by jculvey — APPROVED** [review]

**Pull request merged**
